File: python/medium/spiral_traverse/spiral_traverse.py
# O(n) time | O(n) space, where n is the total number of elements
# in the two-dimensional array
def spiralTraverse(array):
    # Initialize an empty list to store the result
    result = []

    # Define the boundaries for the rows and columns
    startRow, endRow = 0, len(array) - 1
    startCol, endCol = 0, len(array[0]) - 1

    # Loop until all of the boundaries have been crossed
    while startRow <= endRow and startCol <= endCol:
        # Traverse from left to right across the current top row
        for col in range(startCol, endCol + 1):
            result.append(array[startRow][col])  # Add the current element to result

        # Traverse from top to bottom down the current right-most column
        for row in range(startRow + 1, endRow + 1):
            result.append(array[row][endCol])  # Add the current element to result

        # Traverse from right to left across the current bottom row
        for col in reversed(range(startCol, endCol)):
            # Break if there is only one row left to prevent duplication
            if startRow == endRow:
                break
            result.append(array[endRow][col])  # Add the current element to result

        # Traverse from bottom to top up the current left-most column
        for row in reversed(range(startRow + 1, endRow)):
            # Break if there is only one column left to prevent duplication
            if startCol == endCol:
                break
            result.append(array[row][startCol])  # Add the current element to result

        # Move the boundaries inward for the next layer
        startRow += 1  # Move the startRow down
        endRow -= 1  # Move the endRow up
        startCol += 1  # Move the startCol right
        endCol -= 1  # Move the endCol left

    return result

File: java/medium/spiral_traverse/SpiralTraverse.java
package main.java.com.example.spiraltraverse;

import java.util.ArrayList;
import java.util.List;

public class SpiralTraverse {
	// O(n) time | O(n) space, where n is the total number of elements in the two-dimensional array
	public static List<Integer> spiralTraverse(int[][] array) {
		// Check if the input array is empty; if so, return an empty list
		if(array.length == 0) {
			return new ArrayList<Integer>();
		}

		// Initialize a list to store the result of the spiral traversal
		List<Integer> result = new ArrayList<Integer>();

		// Define the starting and ending row and column indices
		int startRow = 0;                   // Starting row index
		int endRow = array.length - 1;      // Ending row index (last row)
		int startCol = 0;                   // Starting column index
		int endCol = array[0].length - 1;   // Ending column index (last column)

		// Continue until all rows and columns are traversed
		while(startRow <= endRow && startCol <= endCol) {
			// Traverse from left to right along the current startRow
			for (int col = startCol; col <= endCol; col++) {
				result.add(array[startRow][col]);  // Add each element to the result
			}

			// Traverse from top to bottom along the current endCol
			for (int row = startRow + 1; row <= endRow; row++) {
				result.add(array[row][endCol]);  // Add each element to the result
			}

			// Traverse from right to left along the current endRow, if there is still a row to process
			for (int col = endCol - 1; col >= startCol; col--) {
				// Break if there is only one row left to prevent duplication
				if(startRow == endRow) {
					break;
				}
				result.add(array[endRow][col]);  // Add each element to the result
			}

			// Traverse from bottom to top along the current startCol, if there is still a column to process
			for (int row = endRow - 1; row > startRow; row--) {
				// Break if there is only one column left to prevent duplication
				if(startCol == endCol) {
					break;
				}
				result.add(array[row][startCol]);  // Add each element to the result
			}

			// Move the boundaries inward for the next layer of the spiral
			startRow++;  // Move the startRow down
			endRow--;    // Move the endRow up
			startCol++;  // Move the startCol right
			endCol--;    // Move the endCol left
		}

		// Return the result containing the elements in spiral order
		return result;
	}
}